keps/sig-cluster-lifecycle/0015-kubeadm-join-master.md

```diff
@@ -0,0 +1,431 @@
+should copy control plane certificates from an existing master node, e.g. bootstrap master
+
+> NB. kubeadm is limited to execute actions *only*
+> in the machine where it is running, so it is not possible to copy automatically
+> certificates from remote locations.
+
+Please note that strictly speaking only ca certificate and and service account key pair
+certificate are required to be equal among all masters. Accordingly:
+
+- `kubeadm join --master` will check for the mandatory certificates and fail fast if
+  they are missing
+- If some/all of the other certificates are provided by the user as well, `kubeadm join --master`
+  will use them without further checks.
+- If any other certificates are missing, `kubeadm join --master` will create them.
+
+> see "Strategies for distributing cluster certificates" paragraph for
+> additional info about this step.
+
+#### The kubeadm join --master workflow
+
+The `kubeadm join --master` workflow will be implemented as an extension of the
+existing `kubeadm join` flow.
+
+`kubeadm join --master` will accept an additional parameter, that is the apiserver advertise
+address of the joining node; as details in following paragraphs, the value assigned to
+this parameter depends on the user choice between a dynamic bootstrap workflow or a static
+bootstrap workflow.
+
+The updated join workflow will be the following:
+
+1. Discovery cluster info [No changes to this step]
+
+   > NB This step waits for a first instance of the kube-apiserver to become ready
+   > (the bootstrap master); And thus it acts as embedded mechanism for handling the sequence
+   > `kubeadm init` and `kubeadm join` actions in case of parallel node creation.
+
+2. In case of `join --master` [New step]
+
+   1. Using the bootstrap token as identity, read the `kubeadm-config` configMap
+      in `kube-system` namespace.
+
+      > This requires to grant access to the above configMap for
+      > `system:bootstrappers` group.
+
+   2. Check if the cluster is ready for joining a new master node:
+
+      a. Check if the cluster has a stable `controlplaneAddress`
+      a. Check if the cluster uses an external etcd
+      a. Checks if the mandatory certificates exists on the file system
+
+   3. Prepare the node for joining as a master node:
+
+      a. Create missing certificates (in any).
+         > please note that by creating missing certificates kubeadm can adapt seamlessly
+         > to a dynamic workflow or to a static workflow (and to apiserver advertise address
+         > of the joining node). see following paragraphs for more details for additional info.
+
+      a. In case of control plane deployed as static pods, create related kubeconfig files
+      and static pod manifests.
+
+      > see "Strategies for deploying control plane components" paragraph
+      > for additional info about this step.
+
+   4. Create the admin.conf kubeconfig file
+
+3. Executes the kubelet TLS bootstrap process [No changes to this step]:
```

Comment:
Why doesn't this happen as step 2 instead of 3?
I think having this as step 2 and making everything `--master` post-node join makes a lot of sense (e.g. in case the normal node join fails)